caledonia/src/bounded.rs

```diff
@@ -0,0 +1,451 @@
+            return false;
+        }
+        let r0 = Round::new(proof.r, proof.d, setup.n_p);
+        let (b, round) = proof.items.iter().fold((true, r0), |(b, r), &s| {
+            (
+                b && r.h_usize == Proof::h0(setup, proof.r, s),
+                Round::update(&r, s),
+            )
+        });
+        return b && Proof::h2(setup, &round);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use rand_chacha::ChaCha20Rng;
+    use rand_core::{RngCore, SeedableRng};
+
+    #[test]
+    fn test_verify() {
+        let mut rng = ChaCha20Rng::from_seed(Default::default());
+        let nb_tests = 1_000;
+        for _t in 0..nb_tests {
+            let seed = rng.next_u32().to_ne_bytes().to_vec();
+            let s_p = utils::gen_items(seed, 100);
+            let params = Params {
+                lambda_sec: 10,
+                lambda_rel: 10,
+                n_p: 80,
+                n_f: 20,
+            };
+            let setup = Setup::new(&params);
+            let proof = Proof::prove(&setup, &s_p);
+            assert!(Proof::verify(&setup, proof.clone()));
+            let proof_0 = Proof {
+                r: proof.r,
+                d: 0,
+                items: proof.items.clone(),
+            };
+            assert!(!Proof::verify(&setup, proof_0));
+            let proof_d = Proof {
+                r: proof.r,
+                d: proof.d.wrapping_add(1),
+                items: proof.items.clone(),
+            };
+            assert!(!Proof::verify(&setup, proof_d));
+            let proof_r = Proof {
+                r: proof.r.wrapping_add(1),
+                d: proof.d,
+                items: proof.items.clone(),
+            };
+            assert!(!Proof::verify(&setup, proof_r));
+            let proof_item = Proof {
+                r: proof.r,
+                d: proof.d,
+                items: Vec::new(),
+            };
+            assert!(!Proof::verify(&setup, proof_item));
+            let mut wrong_items = proof.items.clone();
+            let last_item = wrong_items.pop().unwrap();
+            let mut penultimate_item = wrong_items.pop().unwrap();
+            let proof_itembis = Proof {
+                r: proof.r,
+                d: proof.d,
+                items: wrong_items.clone(),
+            };
+            assert!(!Proof::verify(&setup, proof_itembis));
+            // Modifying the penultimate item to check correctness of H1 check and not H2
+            penultimate_item[0] = penultimate_item[0].wrapping_add(42u8);
+            wrong_items.push(penultimate_item);
+            wrong_items.push(last_item);
+            let proof_itembis = Proof {
+                r: proof.r,
+                d: proof.d,
+                items: wrong_items.clone(),
+            };
+            assert!(!Proof::verify(&setup, proof_itembis));
+        }
+    }
+
+    #[test]
+    fn test_prove() {
```

Comment:
Is this really a `test`? Seems like we print stuff at the end too

Reply:
You are correct. This was to get prelim. benches before using criterion (c.f. PR https://github.com/cardano-scaling/alba/pull/15).